src/common/mongoose/plugins/history.plugin.ts

```diff
@@ -0,0 +1,377 @@
+import { Schema, Document, QueryWithHelpers } from "mongoose"; // Import QueryWithHelpers
+import {
+  GenericHistory,
+  GenericHistoryDocument,
+} from "../../schemas/generic-history.schema";
+import { computeDeltaWithOriginals } from "../../utils/delta.util";
+
+interface HistoryPluginOptions {
+  historyModelName?: string;
+  modelName?: string;
+  trackables?: string[];
+  getOriginator?: () => string | undefined; // Function to get current user context if needed
+}
+
+// Define an interface for the query context including our custom property
+// Make it generic over the Document type (T) and the Result type (Res)
+// Use QueryWithHelpers for better compatibility with Mongoose's internal types
+interface QueryWithHistory<
+  Res, // Result type of the query (e.g., T | null, DeleteResult)
+  T extends Document, // Document type
+> extends QueryWithHelpers<Res, T, object> {
+  // Use object for THelpers
+  _originalDoc?: T | (T & Document<unknown, object, T>[]) | null; // Store the original doc(s) fetched
+}
+
+export function historyPlugin(
+  schema: Schema,
+  options: HistoryPluginOptions = {},
+) {
+  const {
+    historyModelName = GenericHistory.name,
+    modelName: optionsModelName, // Extract modelName if provided in options
+    trackables = (process.env.TRACKABLES?.split(",") || []).map((t) =>
+      t.trim(),
+    ),
+    getOriginator,
+  } = options;
+
+  const trackableStrategy =
+    process.env.TRACKABLE_STRATEGY?.toLowerCase() === "delta"
+      ? "delta"
+      : "document";
+
+  // Get the model name more robustly - try multiple approaches
+  let modelName: string | undefined;
+
+  // First try: use modelName from options if provided
+  if (optionsModelName) {
+    modelName = optionsModelName;
+  }
+  // Second try: try to access collection name through various methods
+  else {
+    try {
+      // Use type assertion to access options property safely
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      const schemaOptions = (schema as any).options;
+      if (schemaOptions && schemaOptions.collection) {
+        modelName = schemaOptions.collection;
+      }
+    } catch (e) {
+      console.warn("Error accessing schema options:", e);
+    }
+
+    // Third try: schema.get("collection") - this might be failing too
+    if (!modelName && typeof schema.get === "function") {
+      try {
+        modelName = schema.get("collection");
+      } catch (e) {
+        console.warn("Error getting collection from schema.get:", e);
+      }
+    }
+
+    // Fourth try: check if schema has a constructor with a modelName property
+    if (!modelName) {
+      try {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        const constructor = (schema as any).constructor;
+        if (constructor && constructor.modelName) {
+          modelName = constructor.modelName;
+        }
+      } catch (e) {
+        console.warn("Error accessing schema constructor:", e);
+      }
+    }
+  }
```

Comment:
Do we need all this logic? ...or can we simplify and maybe require the model name